versioned/persist/mongodb/src/main/java/org/projectnessie/versioned/persist/mongodb/MongoDatabaseAdapter.java

```diff
@@ -0,0 +1,401 @@
+  }
+
+  private void delete(MongoCollection<Document> collection, Set<Hash> ids) {
+    DeleteResult result = collection.deleteMany(Filters.in("_id", toIds(ids)));
+
+    if (!result.wasAcknowledged()) {
+      throw new IllegalStateException("Unacknowledged write to " + collection.getNamespace());
+    }
+  }
+
+  private byte[] loadById(MongoCollection<Document> collection, String id) {
+    Document doc = collection.find(Filters.eq(id)).first();
+    if (doc == null) {
+      return null;
+    }
+
+    Binary data = doc.get("data", Binary.class);
+    if (data == null) {
+      return null;
+    }
+
+    return data.getData();
+  }
+
+  private <T> T loadById(MongoCollection<Document> collection, Hash id, Parser<T> parser) {
+    return loadById(collection, toId(id), parser);
+  }
+
+  private <T> T loadById(MongoCollection<Document> collection, String id, Parser<T> parser) {
+    byte[] data = loadById(collection, id);
+    if (data == null) {
+      return null;
+    }
+
+    try {
+      return parser.parse(data);
+    } catch (InvalidProtocolBufferException e) {
+      throw new IllegalStateException(e);
+    }
+  }
+
+  /**
+   * Fetches a collection of documents by hash and returns them in the order of hashes requested.
+   *
+   * <p>Note: unknown hashes will correspond to {@code null} elements in the result list.
+   */
+  private <T> List<T> fetchMappedPage(
+      MongoCollection<Document> collection, List<Hash> hashes, Function<Document, T> mapper) {
+    List<String> ids = hashes.stream().map(this::toId).collect(Collectors.toList());
+    FindIterable<Document> docs = collection.find(Filters.in("_id", ids)).limit(hashes.size());
+
+    HashMap<Hash, Document> loaded = new HashMap<>(hashes.size());
+    for (Document doc : docs) {
+      loaded.put(idAsHash(doc), doc);
+    }
+
+    List<T> result = new ArrayList<>(hashes.size());
+    for (Hash hash : hashes) {
+      T element = null;
+      Document document = loaded.get(hash);
+      if (document != null) {
+        element = mapper.apply(document);
+      }
+
+      result.add(element); // nulls elements are permitted
+    }
+
+    return result;
+  }
+
+  private <T> List<T> fetchPage(
+      MongoCollection<Document> collection, List<Hash> hashes, Parser<T> parser) {
+    return fetchMappedPage(
+        collection,
+        hashes,
+        document -> {
+          try {
+            byte[] data = data(document);
+            return parser.parse(data);
+          } catch (InvalidProtocolBufferException e) {
+            throw new IllegalStateException(e);
+          }
+        });
+  }
+
+  @Override
+  protected CommitLogEntry fetchFromCommitLog(NonTransactionalOperationContext ctx, Hash hash) {
+    return loadById(client.getCommitLog(), hash, ProtoSerialization::protoToCommitLogEntry);
+  }
+
+  private Hash idAsHash(Document doc) {
+    String id = doc.get("_id", String.class);
+    String hash = id.substring(keyPrefix.length() + 1); // + 1 for the separator char
```

Comment:
Better use String.indexOf() here instead of assuming the same key-prefix (unlikely that this might happen)

Reply:
I've added some validation in e6b4f6eebea447420690ab610184702adf28f57c

Reply:
That's a lot ;)
I'm okay with that.